pkg/controller/servicebinding/secret.go

```diff
@@ -27,10 +32,68 @@ func (s *secret) buildResourceClient() dynamic.ResourceInterface {
 	return s.client.Resource(gvr).Namespace(s.ns)
 }
 
-// createOrUpdate will take informed payload and either create a new secret or update an existing
-// one. It can return error when Kubernetes client does.
-func (s *secret) createOrUpdate(payload map[string][]byte, ownerReference metav1.OwnerReference) (*unstructured.Unstructured, error) {
+// compare existing secret and new secret
+func (s *secret) isSame(secretName string, payload map[string][]byte) bool {
 	logger := s.logger.WithValues("Namespace", s.ns, "Name", s.name)
+
+	resourceClient := s.buildResourceClient()
+	existingSecret, err := resourceClient.Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		logger.Error(err, "Error fetching secret")
+		return false
+	}
+	existingSecretData, _, _ := unstructured.NestedMap(existingSecret.Object, "data")
+
+	payloadInterim := make(map[string]interface{})
+	for k, v := range payload {
+		payloadInterim[k] = base64.StdEncoding.EncodeToString(v)
+	}
+
+	comparisonResult := nestedMapComparison(existingSecretData, payloadInterim)
+
+	if comparisonResult.Success {
+		logger.Debug("Secret data is same.")
+		return true
+	}
+
+	return false
+}
+
+func sortKeys(payload map[string][]byte) string {
+	keys := make([]string, 0, len(payload))
+	for k := range payload {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	var text []string
+	for _, k := range keys {
+		s := fmt.Sprintf("%s%s", k, string(payload[k]))
+		text = append(text, s)
+	}
+	return strings.Join(text, "")
+}
+
+func hash(text string) (string, error) {
+	algorithm := sha1.New()
+	_, err := algorithm.Write([]byte(text))
+	if err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(algorithm.Sum(nil)), nil
+}
+
+// create will take informed payload and create a new secret
+// It can return error when Kubernetes client does.
+func (s *secret) create(payload map[string][]byte, ownerReference metav1.OwnerReference) (*unstructured.Unstructured, error) {
+	logger := s.logger.WithValues("Namespace", s.ns, "Name", s.name)
+
+	uniqueSuffix, err := hash(sortKeys(payload))
+	if err != nil {
+		return nil, err
```

Comment:
Wrap the error with useful information for `create` callers to inspect; passing  through `err` makes it hard to identify where errors come from since there's no context associated with the error.

Reply:
Is this okay? - f1b2c79

Reply:
That's good to have the log, but I was alluding to `errors.Wrap`.

The main reason is that the error returned in `HashedResourceNameForData` comes from `hash.Write`; if this error reaches the controller as-is, there's no clue other than inspecting the string to find the error so we should always wrap the error (either a string or a specific test type) to make it easier to locate the error source.

Reply:
Okay. I will add a fix for that.

Reply:
Done here https://github.com/redhat-developer/service-binding-operator/pull/750/commits/bb98f74ab2f634a28526c0f290b1248847c722d6